### DISCUSSION TITLE
Association Proxy for one-to-one relationships

### DISCUSSION THREAD

**Discussion by miketheman** [sqlalchemy/sqlalchemy#10507] [Usage Questions] [answered]
Hi there!

I've been poking at this problem for a little while, and am thoroughly stumped - I'm probably missing something subtle.

What I'm trying to do:
- Have two models (user, system) be members of a generic association table (observer)
- (Eventually) have another set of models related to the id in the association table

Said another way, either a User, System (and any future model) mix in a relationship to the Observer table with a discriminator. In my head Observer looks like this:

| id | discriminator | parent_id |
| --- | --- | --- |
| 1 | user | 1234 |
| 2 | system | 5678 |

And then Observations would only store a reference to `Observer.id`.

After reading through a bunch of the relationship docs, it seems like the closest approach would be [discriminator_on_association](https://docs.sqlalchemy.org/en/20/_modules/examples/generic_associations/discriminator_on_association.html), which uses `association_proxy`.

The example works for a one-to-many relationship between a Customer/Supplier and any Addresses, but I'm trying to make a one-to-one relationship between User/System and Observer record, not a list.

I updated the example code to express the single-ness I want, but fails with:

`TypeError: Incompatible collection type: Address is not list-like`

Association Proxy [docstring says it works for one-to-one](https://github.com/sqlalchemy/sqlalchemy/blob/efe4025c068d6746637db362245ab88ce7dfe785/lib/sqlalchemy/ext/associationproxy.py#L108-L111), but I guess I'm just not "getting it" on what else needs to change to make this work.

Updated example that fails (comments removed). Assumption is a given Company or Supplier will only a single Address.

```python
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.declarative import as_declarative
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import backref
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


@as_declarative()
class Base:
    @declared_attr
    def __tablename__(cls):
        return cls.__name__.lower()

    id = Column(Integer, primary_key=True)


class AddressAssociation(Base):
    __tablename__ = "address_association"

    discriminator = Column(String)
    """Refers to the type of parent."""

    __mapper_args__ = {"polymorphic_on": discriminator}


class Address(Base):
    association_id = Column(Integer, ForeignKey("address_association.id"))
    street = Column(String)
    city = Column(String)
    zip = Column(String)
    association = relationship("AddressAssociation", backref="addresses")

    parent = association_proxy("association", "parent")

    def __repr__(self):
        return "%s(street=%r, city=%r, zip=%r)" % (
            self.__class__.__name__,
            self.street,
            self.city,
            self.zip,
        )


class HasAddresses:
    @declared_attr
    def address_association_id(cls):
        return Column(Integer, ForeignKey("address_association.id"))

    @declared_attr
    def address_association(cls):
        name = cls.__name__
        discriminator = name.lower()

        assoc_cls = type(
            "%sAddressAssociation" % name,
            (AddressAssociation,),
            dict(
                __tablename__=None,
                __mapper_args__={"polymorphic_identity": discriminator},
            ),
        )

        cls.address = association_proxy(
            "address_association",
            "addresses",
            creator=lambda addresses: assoc_cls(addresses=addresses),
        )
        return relationship(
            assoc_cls, backref=backref("parent", uselist=False)
        )


class Customer(HasAddresses, Base):
    name = Column(String)


class Supplier(HasAddresses, Base):
    company_name = Column(String)


engine = create_engine("sqlite://", echo=True)
Base.metadata.create_all(engine)

session = Session(engine)

session.add_all(
    [
        Customer(
            name="customer 1",
            address=Address(
                street="123 anywhere street", city="New York", zip="10110"
            ),
        ),
        Supplier(
            company_name="Ace Hammers",
            addresses=Address(street="2569 west elm", city="Detroit", zip="56785")
        ),
    ]
)

session.commit()

for customer in session.query(Customer):
    print(customer.address)
    print(customer.address.parent)
```

**Comment by zzzeek** [comment, marked as answer]
you're missing a uselist=False on Address:

```
class Address(Base):
    association_id = Column(Integer, ForeignKey("address_association.id"))
    street = Column(String)
    city = Column(String)
    zip = Column(String)
    association = relationship("AddressAssociation", backref=backref("addresses", uselist=False))

```

and also sent the wrong keyword name to Supplier:

```
session.add_all(
    [
        Customer(
            name="customer 1",
            address=Address(
                street="123 anywhere street", city="New York", zip="10110"
            ),
        ),
        Supplier(
            company_name="Ace Hammers",
            address=Address(
                street="2569 west elm", city="Detroit", zip="56785"
            ),
        ),
    ]
)

```


it may be more verbose but the code here might be easier to follow if it used [back_populates](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.back_populates) instead of backref.   this mapping is an example of why we've moved "backref" into "legacy" status.

**Reply by miketheman** [reply]
Thanks, I suspected it was something rather small!
And I never got to Supplier to see that issue, thanks!

Now that the example is working, I'll poke at making it work with back_populates instead. 

I'd be happy to submit a PR for an updated declarative annotated version of the example if that's helpful.

**Reply by miketheman** [reply]
I submitted https://github.com/sqlalchemy/sqlalchemy/pull/10450 - I hope that's correct.

**Reply by zzzeek** [reply]
this is a PR you submitted....before you opened this discussion in the first place, is that right?    it's got a lower number and is dated October 10 whereas this discussion is October 19....so that's just a general PR right?

**Reply by miketheman** [reply]
@zzzeek you're correct - the PR I linked is for table_per_related, not discriminator_on_association - my mistake for conflating them! I've been reading these so much lately, my brain couldn't keep them separate.
